src/Symfony/Component/DependencyInjection/Loader/FileLoader.php

```diff
@@ -56,10 +56,29 @@ public function registerClasses(Definition $prototype, $namespace, $resource, $e
 
         $classes = $this->findClasses($namespace, $resource, $exclude);
         // prepare for deep cloning
-        $prototype = serialize($prototype);
+        $serializedPrototype = serialize($prototype);
+        $interfaces = array();
+        $interfaceImplementations = array();
 
         foreach ($classes as $class) {
-            $this->setDefinition($class, unserialize($prototype));
+            if (interface_exists($class, false)) {
+                $interfaces[] = $class;
+            } else {
+                $this->setDefinition($class, unserialize($serializedPrototype));
+                foreach (class_implements($class, false) as $interface) {
+                    $interfaceImplementations[$interface][] = $class;
+                }
+            }
+        }
+        foreach ($interfaces as $interface) {
+            if (!isset($interfaceImplementations[$interface]) || isset($interfaceImplementations[$interface][1])) {
+                continue;
+            }
+            foreach ($interfaceImplementations[$interface] as $class) {
```

Comment:
The `foreach` looks a bit funny. At this point, we know there is only *one* class that implements this interface, right? So this code is identical to *not* having the `foreach` and using `$interfaceImplementations[$interface][0]`, right? If so, the `foreach` just reads a bit confusing

Reply:
was here to be sure someone actually read the code :P
removed thanks